control change notification handling

ujfalusi opened the pull request:
If the value/data associated with a control changes in SOF it will send a
notification (SOF_IPC_GLB_COMP_MSG with SOF_IPC_COMP_GET_VALUE/DATA).

When such notification arrives we use snd_ctl_notify_one() to signal
userspace about the change.

The kernel is prepared for two types of notification:
- the notification carries the new data for the control (num_elems != 0)
The new value/data is copied to the control's local data

- blank message about a change
The new flag for the scontrol (updated_in_firmware) is set and when next
time user space reads the value via the kcontrol's get callback we will
refresh the control's local data from the firmware.

The patches has been tested by @singalsu using the beamformer code in SOF, alsamixer updates the displayed values on change.
While the TDFB is using enum type, the kernel code can support notification from all supported kcontrol type.

ujfalusi commented:
changes since v1:
- make sure that malformed message (num_elems > 0) with no actual data in the message is treated as empty message
- fix sparse errors and also add back the missing code to snd_sof_refresh_control() to set the ctrl_type and ctrl_cmd.

singalsu commented on sound/soc/sof/control.c:
```diff
@@ -511,3 +568,109 @@ int snd_sof_bytes_ext_get(struct snd_kcontrol *kcontrol,
+
+	if (cdata->rhdr.hdr.size == sizeof(struct sof_ipc_ctrl_data))
+		/* the message does not include additional data */
+		cdata->num_elems = 0;
```
Would it be better to ignore message and error to kernel log if size and num_elems does not match?

ujfalusi replied:
@singalsu, you are right, let's ignore such notification.

Along the same line: I might revisit the debug print levels, I tried the amount relaxed, but I might be able to remove or downgrade some.

plbossart commented on sound/soc/soc-topology.c:
```diff
@@ -1478,8 +1478,7 @@ static int soc_tplg_dapm_widget_create(struct soc_tplg *tplg,
 	control_hdr = (struct snd_soc_tplg_ctl_hdr *)tplg->pos;
-	dev_dbg(tplg->dev, "ASoC: template %s has %d controls of type %x\n",
-		w->name, w->num_kcontrols, control_hdr->type);
+	dev_dbg(tplg->dev, "ASoC: template %s has %d controls\n", w->name, w->num_kcontrols);
```
Rather than suppress all information on type, can we instead loop to show what types are enabled for that control?

ujfalusi replied:
But then we shall have:
```
for (i = 0; i < w->num_kcontrols; i++) {
	dev_dbg();
	// or
	sprintf(dbg_buff, "");
}
```
I would rather leave that out, but I can also just drop this patch, it just occurred to me that this is now wrong.

ranj063 replied:
you dont need the loop, but maybe we can move the dev_dbg() to after creating the controls and use mixer_coutn/enum_count/bytes_count to print instead?

ujfalusi replied:
@ranj063, you are right, I have moved it after the loop.

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -79,6 +79,8 @@ struct snd_sof_control {
 
 	struct snd_sof_led_control led_ctl;
+
+	bool updated_in_firmware;
```
does this mean volatile?

I am struggling to understand what this means in terms of access.

ujfalusi replied:
Not volatile, but the value/data of the control can change in SOF and the kernel should sync it's cached value to reflect the current state.

ranj063 replied:
this flag would make the kcontrol get behaviour equivalent to a volatile get. I'd like a different name more indicative name something like "control_data_dirty" maybe?

ujfalusi replied:
In a way, yes, it makes it to do an IPC like the volatile does if the flag is set, but this flag will not make the control similar to the volatile type we currently have.
If the firmware does not send event for a control (ever) then this flag will not change the behavior.

I would choose something different than `control_data_dirty`, the control have value(s) or data, the code is specific about that.
A simple `dirty` perhaps? With a comment that the firmware's and kernel's value/data is out of sync and need to be refreshed. Or perhaps `changed_in_firmware` or `ipc_refresh` ?

ranj063 replied:
just dirty seems too simple. elem_value_dirty? or simple elem_dirty?

ujfalusi replied:
I'll name it after the IPC wrapper function we use to get the value/data: `snd_sof_ipc_set_get_comp_data()`
Let's try: `comp_data_dirty` ?

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -366,6 +366,36 @@ void snd_sof_ipc_reply(struct snd_sof_dev *sdev, u32 msg_id)
+		/* read back full message */
+		ret = snd_sof_ipc_msg_data(sdev, NULL, cdata, hdr->size);
+		if (ret < 0) {
+			dev_err(sdev->dev, "failed to read component event: %d\n", ret);
```
error: failed to stay with the existing conventions.

ujfalusi replied:
Sorry, I know the convention. I just missed this for some reason.

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -366,6 +366,36 @@ void snd_sof_ipc_reply(struct snd_sof_dev *sdev, u32 msg_id)
+		/* read back full message */
+		ret = snd_sof_ipc_msg_data(sdev, NULL, cdata, hdr->size);
+		if (ret < 0) {
+			dev_err(sdev->dev, "failed to read component event: %d\n", ret);
```
if a component can have more that one control, how do we know which of the controls received a notification?
Shouldn't there be something that identifies which control and control type received IPC?

ujfalusi replied:
We do that, we look up the swidget with the matching comp_id, then we identify the matching kcontrol based on the type and index, but it is done in the snd_sof_control_notify()
Here we just fetch the notification message.

plbossart commented on sound/soc/sof/control.c:
```diff
@@ -65,6 +65,53 @@ static inline u32 ipc_to_mixer(u32 value, u32 *volume_map, int size)
+				    "error: %s pm_runtime_get_sync failed: %d\n",
+				    __func__, ret);
+		pm_runtime_put_noidle(scomp->dev);
+		return;
```
can it happen that you have to resume the device after getting a firmware notification?

I don't fully get the flow here, on a notification userspace does a get and this is to protect against the device suspending in the mean time? If yes, what are the odds that the value is still relevant?

ujfalusi replied:
I have been hesitant around this.
Most likely we should just:
```
if (!pm_runtime_active(scomp->dev))
	return;
```
But if this is an issue then we already have it in use via `snd_sof_bytes_ext_volatile_get()`, my thinking was that since that code is already there, let's do the same here.

plbossart commented on sound/soc/sof/control.c:
```diff
@@ -511,3 +568,109 @@ int snd_sof_bytes_ext_get(struct snd_kcontrol *kcontrol,
+		scontrol->updated_in_firmware = true;
+
+	snd_ctl_notify_one(swidget->scomp->card->snd_card,
+			   SNDRV_CTL_EVENT_MASK_VALUE, kc, 0);
```
can you add a comment on what happens once this notification is done? I don't get the flow at all, thanks!

ujfalusi replied:
It depends. `snd_ctl_notfy()`/`snd_ctl_notify_one()` is just sends an event to userspace _if_ the related devfs control file is open and there is a wait issued on it. It is just wakes up userspace for the event.
From there it is up to userspace, but it will most likely going to read the kcontrol's value/data to see what changed. alsamixer does this for example. If you have it open then the given kcontrol is changing real time.

If there is no one interested about the event then this is a simple NOP.

plbossart reviewed (commented):
thanks @ujfalusi I couldn't get the full idea so logged my review comments to help clarify this PR.

ujfalusi commented:
> thanks @ujfalusi I couldn't get the full idea so logged my review comments to help clarify this PR.

The idea is that SOF will be able to send notification when a control's value/data has changed _within_ the firmware because of internal processing or events.
We have support for binary volatile control type, but we might have features where enum/switch/volume changes. Re-implementing everything as volatile as well would be not much of a gain for several reasons:
volatile controls would do an IPC all the time, regardless if there is a need or not.
We still don't have notification which forces userspace to continuously poll.

With this series (the last patch actually) we support a notification from SOF and we only reach out to DSP to fetch the new value/data when there is a need. Userspace does not need to poll, just wait for the event to come about the change on the kcontrol.

plbossart commented:
> > thanks @ujfalusi I couldn't get the full idea so logged my review comments to help clarify this PR.
> 
> The idea is that SOF will be able to send notification when a control's value/data has changed _within_ the firmware because of internal processing or events.
> We have support for binary volatile control type, but we might have features where enum/switch/volume changes. Re-implementing everything as volatile as well would be not much of a gain for several reasons:
> volatile controls would do an IPC all the time, regardless if there is a need or not.
> We still don't have notification which forces userspace to continuously poll.
> 
> With this series (the last patch actually) we support a notification from SOF and we only reach out to DSP to fetch the new value/data when there is a need. Userspace does not need to poll, just wait for the event to come about the change on the kcontrol.

Thanks, you should probably add this flow to the commit message so that folks like me who never understood the control layer get an idea. 

On the pm_runtime question, presumably if you get a notification, can you not just do the cache update then, instead of doing it later on a get? Presumably only when the component is active and hence the DSP is active does the notification show up.
That way on the second part where userspace does a get all the IPC to retrieve the information is done, and even if the DSP went to sleep again userspace would get the last cached value.

ujfalusi commented:
> Thanks, you should probably add this flow to the commit message so that folks like me who never understood the control layer get an idea.

Sure, I will update the commit message, thanks for the suggestion!
 
> On the pm_runtime question, presumably if you get a notification, can you not just do the cache update then, instead of doing it later on a get? Presumably only when the component is active and hence the DSP is active does the notification show up.
> That way on the second part where userspace does a get all the IPC to retrieve the information is done, and even if the DSP went to sleep again userspace would get the last cached value.

We can do an immediate readback as the notification processing is in interrupt context (msg_rx path) and we get ipc_timeout - it was the first thing I have tried.
I would say that SOF will only send notification if it is active and userspace will only trigger a GET if it cares, so all the things will happen while we are active.
You are right, if the DSP powers off between the notification and the GET then we would loose context and the readback would not retrieve valid data, so I will switch to the runtime check only, what do you think?

plbossart commented:
@lyakh don't we have a limitation that we cannot initiate an IPC TX while dealing with an IPC RX? If yes, we'd need to have a workqueue deal with the notification at a later time - which could possibly be racy if the userpace does a GET while the notification workqueue is pending.

ujfalusi commented:
> @lyakh don't we have a limitation that we cannot initiate an IPC TX while dealing with an IPC RX? If yes, we'd need to have a workqueue deal with the notification at a later time - which could possibly be racy if the userpace does a GET while the notification workqueue is pending.

Yes, we have that limitation. It has never been a requirement to be able to do IPC TX right in the IPC RX processing.
Since this change notification feature is aimed for user space, this is not a problem:
- if user space is waiting on event then it will do an update for the kcontrol
- if no one is interested in a change, then it does not matter if the kcontrol's value is not updated as no one is checking it. But when someone comes and checks it then we will get the fresh value/data from DSP.

If at some point we want to have SOF triggered DAPM changes then we can come back and address it, but it is not a target of this series in any ways.

cujomalainey commented:
@chiang831 FYI

ranj063 commented on sound/soc/sof/control.c:
```diff
@@ -244,6 +297,8 @@ int snd_sof_bytes_get(struct snd_kcontrol *kcontrol,
 	struct sof_abi_hdr *data = cdata->data;
 	size_t size;
 
+	snd_sof_refresh_control(scontrol);
```
this makes sense for bytes controls, but volume/enum/switch controls too? Why will the firmware automatically mess with these values?

juimonen replied:
We we're using enum with @singalsu for direction-of-arrival "angles" callback with beamformer. Could be done with bytes, but would be totally proprietary (unreadable) by normal people.

Anyway despite the above, to me it would seem appropriate just to refresh all control types. Can't imagine what harm it could do and we would be prepared for all possible future scenarios.

ujfalusi replied:
@ranj063, I agree with @juimonen that having a generic implementation is better so in the future when SOF gains more autonomous features or adds support for different changes benefiting the user then we do not have to revisit the code.

I'm not sure how user space handles the current SOF_TPLG_KCTL_BYTES_VOLATILE_RO, but if they need to poll to notice a change then they can be switch to this notification method without the volatile mode.
Even if user space keeps polling the firmware and topology can be changed to take advantage of the fact that we no longer need to do IPC if the control is not changed on the SOF side..

singalsu replied:
I think it's nice that this works for all controls. The existing basic components will remain kernel to FW direction. But we could have instead of fixed volume control an automatic gain control or level control with user (initial) gain setting. The gain changes done by algorithm would be visible to user space too. Switch is practical for binary audio events indication. I think this opens the space for a lot of innovation.

singalsu replied:
@ranj063 There's platform power impact if the updates happen often so need to be careful and do internal filtering in FW to have e.g. thresholds and hysteresis to pass them to user space. The recommendation is that a component should have switch control (default off) to enable updates. I'm adding some examples to beamformer PR to show FW code and topology example.

lyakh commented on sound/soc/sof/ipc.c:
```diff
@@ -366,6 +366,36 @@ void snd_sof_ipc_reply(struct snd_sof_dev *sdev, u32 msg_id)
+	switch (msg_type) {
+	case SOF_IPC_COMP_GET_VALUE:
+	case SOF_IPC_COMP_GET_DATA:
+		cdata = kzalloc(hdr->size, GFP_KERNEL);
```
does it really have to be zero-initialised?

ujfalusi replied:
Not really, I'll switch it to kmalloc

lyakh commented on sound/soc/sof/ipc.c:
```diff
@@ -366,6 +366,36 @@ void snd_sof_ipc_reply(struct snd_sof_dev *sdev, u32 msg_id)
+		ret = snd_sof_ipc_msg_data(sdev, NULL, cdata, hdr->size);
+		if (ret < 0) {
+			dev_err(sdev->dev, "failed to read component event: %d\n", ret);
+			return;
```
this looks like a memory leak

ujfalusi replied:
You are right, it is, thanks.

lyakh commented on sound/soc/sof/ipc.c:
```diff
@@ -366,6 +366,36 @@ void snd_sof_ipc_reply(struct snd_sof_dev *sdev, u32 msg_id)
+		}
+		break;
+	default:
+		dev_err(sdev->dev, "error: unhandled component message %x\n", msg_type);
```
was it `%#x` to include a `0x` prefix?

ujfalusi replied:
No, it does not.
I have based the skeleton on the ipc_trace_message() / ipc_stream_message(), the `%x` is coming from there.

ujfalusi commented:
Version 3 is pushed, all comments were addressed:
- commit message is updated for the last patch
- "error:" prefix for all dev_err() prints added and 0x prefix for %x
- Only check if pm_runtime is active  in snd_sof_refresh_control()
- Use kmalloc for cdata allocation and fix memory leak
- Move the dev_dbg() and print the number of control types per template in patch 2

New patch to add 0x prefixes to dev_err() prints in ipc.c

kv2019i commented on sound/soc/sof/ipc.c:
```diff
@@ -405,7 +438,9 @@ void snd_sof_ipc_msgs_rx(struct snd_sof_dev *sdev)
 	case SOF_IPC_GLB_PM_MSG:
+		break;
 	case SOF_IPC_GLB_COMP_MSG:
+		ipc_comp_notification(sdev, &hdr);
```
@singalsu @ujfalusi Just to double-check, I assume old firmware versions will just not send these? Scenario in my mind was that if some old firmware sent these IPCs and whether kernel would handle messages (from the old FW) correctly. But based on my understanding, this seems fine.

ujfalusi replied:
@kv2019i, yes, that is correct: old and current FW does not send these messages. The kernel prior to this patch would ignore the messages.

kv2019i reviewed (approved):
Very neat, thanks @ujfalusi !

plbossart commented on sound/soc/sof/control.c:
```diff
@@ -65,6 +65,37 @@ static inline u32 ipc_to_mixer(u32 value, u32 *volume_map, int size)
+	if (ret < 0)
+		dev_err(scomp->dev, "error: failed to get control data: %d\n", ret);
+
+	scontrol->updated_in_firmware = false;
```
is there a possibility of a race condition, e.g. with the DSP sending two IPCs. If the refresh_control deals with the first one and clears this flag after it's set by the second IPC, it could be that the second update is lost? the test on line 75 would prevent the second access to the register.

ujfalusi replied:
The chances are pretty low that it would happen, but in theory it is a possibility.
The IPC tx/rx is lock protected so the IPC is serialized, if the whole RX processing happens between snd_sof_ipc_set_get_comp_data() and we clear the flag then we can miss one update.

It can be mitigated by moving the flag clear before snd_sof_ipc_set_get_comp_data(), with that we might force a second read via IPC in the worst case but we will not miss updates.

Thank you for pointing this out!

lyakh commented on sound/soc/sof/ipc.c:
```diff
@@ -532,7 +532,7 @@ static void ipc_stream_message(struct snd_sof_dev *sdev, u32 msg_cmd)
 		break;
 	default:
-		dev_err(sdev->dev, "error: unhandled stream message %x\n",
+		dev_err(sdev->dev, "error: unhandled stream message 0x%x\n",
```
I meant "#%x" as documented e.g. in https://man7.org/linux/man-pages/man3/printf.3.html

ujfalusi replied:
you mean: "%#x", right?

Since I will anyway going to update the PR, I'll make this change as well.

ujfalusi replied:
But do we care about coding style and consistency?
```
git grep -nc "%#x" sound/soc/sof/
sound/soc/sof/debug.c:1
sound/soc/sof/ops.h:2
sound/soc/sof/topology.c:1
```
vs
```
git grep -c -e "0x%x" -e "0x%08x" sound/soc/sof/
sound/soc/sof/control.c:2
sound/soc/sof/imx/imx-common.c:1
sound/soc/sof/imx/imx8.c:2
sound/soc/sof/imx/imx8m.c:2
sound/soc/sof/intel/atom.c:1
sound/soc/sof/intel/bdw.c:5
sound/soc/sof/intel/byt.c:4
sound/soc/sof/intel/cnl.c:2
sound/soc/sof/intel/hda-ctrl.c:8
sound/soc/sof/intel/hda-dsp.c:4
sound/soc/sof/intel/hda-ipc.c:3
sound/soc/sof/intel/hda-stream.c:5
sound/soc/sof/intel/hda.c:2
sound/soc/sof/intel/pci-tng.c:4
sound/soc/sof/ipc.c:11
sound/soc/sof/loader.c:15
sound/soc/sof/pcm.c:2
sound/soc/sof/topology.c:7
sound/soc/sof/xtensa/core.c:1
```
The "0x%x" is more like default in sof.

plbossart replied:
not really @ujfalusi, it was more ignorance that %# even existed rather than a conscious decision to use 0x%x...
I think we should gradually convert all those 0x%x to the recommended notation. all the last patches use 0x%#.

ujfalusi replied:
@plbossart, OK. I'll update the PR in the morning.

lyakh reviewed (approved):
not sure how important "0x%x" vs. "%#x" is

ujfalusi commented:
Changes since v2:
- move the `scontrol->updated_in_firmware = false` prior to snd_sof_ipc_set_get_comp_data() in snd_sof_refresh_control()

ujfalusi commented:
Changes since v3:
- Use `"%#x"` instead `"0x%x"` in error prints as requested by @plbossart

ranj063 commented on sound/soc/sof/control.c:
```diff
@@ -65,6 +65,37 @@ static inline u32 ipc_to_mixer(u32 value, u32 *volume_map, int size)
+	cdata->data->abi = SOF_ABI_VERSION;
+
+	/* refresh the component data from DSP */
+	scontrol->updated_in_firmware = false;
```
shouldnt this be updated only if the ipc below succeeds?

ujfalusi replied:
Good point, let me think about this.
The reason why it is moved here is that if we clear it after the IPC then we would have opened a small window for a race.
On the other hand if the IPC fails then either the firmware is dead or for some reason the we can not read the control's value. In the first case there is not much point of retrying, on the second one how likely the second similar command would succeed?

ujfalusi replied:
I think setting back `updated_in_firmware` to `true` in case of IPC failure should do it without introducing race, but I'm not convinced that it will be beneficial.
User space gets notification, it tries to read the data, we got IPC error, the value is not updated, user space takes the old data as new one and it is happy, not requesting it again.

ranj063 commented on sound/soc/sof/ipc.c:
```diff
@@ -366,6 +366,39 @@ void snd_sof_ipc_reply(struct snd_sof_dev *sdev, u32 msg_id)
+	case SOF_IPC_COMP_GET_DATA:
+		cdata = kmalloc(hdr->size, GFP_KERNEL);
+		if (!cdata)
+			return;
```
should we print -ENOMEM error here?

ujfalusi replied:
OK

ujfalusi replied:
Actually kmalloc should print it.

ranj063 commented on sound/soc/sof/control.c:
```diff
@@ -511,3 +552,131 @@ int snd_sof_bytes_ext_get(struct snd_kcontrol *kcontrol,
+		snd_sof_update_control(scontrol, cdata);
+	else
+		/* Mark the scontrol that the value/data is changed in SOF */
+		scontrol->updated_in_firmware = true;
```
@ujfalusi please help me understand the logic here. When would cdata->numelems be 0? and why will we set the flag only if it is 0?

ujfalusi replied:
If the notification message contains the changed value (num_elems != 0) we read it from the message and update the control's storage, no need to do an IPC.
If the notification is just a notification without the value/data then we will need to do an IPC to get it.
Next time user space will reads the kcontrol's value/data we will do an IPC to get the fresh information. If userspace does not care then we will not do an IPC. It does not matter if the value for the control is not correct as user space does not care about it either.
We use the notification to user space to tell that the value has changed, if you care, read it.

The other constraint is that we can not initiate IPC while we are handling the notification (rx).

Initially it was planned that SOF will only send num_elems == 0, but I thought that it might comes handy if we can receive the data as well.

ranj063 replied:
Thanks @ujfalusi. the comment /* the message includes the updated value/data */ is too concise. Maybe you need to spell out that the kcontrol's cache is updated with the message data and no need to resume the DSp during get again or something along those lines.

ujfalusi replied:
@ranj063, good idea, I will extend the comment section.

ujfalusi commented:
Changes since v4:
- flag name change from `updated_in_firmware` to `comp_data_dirty`
- comment section extended to spell out that we are updating the controls's cache using the received notification
- Set the com_data_dirty back to true if the IPC to get it failed

lyakh commented on sound/soc/sof/control.c:
```diff
@@ -511,3 +555,134 @@ int snd_sof_bytes_ext_get(struct snd_kcontrol *kcontrol,
+		size = min_t(size_t, cdata->num_elems, local_cdata->data->size);
+		memcpy(local_cdata->data, cdata->data, size);
+	} else {
+		count = min_t(int, cdata->num_elems, scontrol->num_channels);
```
for my understanding: is it actually valid in both `min_t()` cases to have those values unequal? What would be such cases?

ujfalusi replied:
valid point, let's ignore if the sizes are different.

ujfalusi commented:
Changes since v5:
- do not copy the value/data in case of num_elems mismatch between the notification and local storage.
